agent/engine/docker_task_engine.go

```diff
@@ -2108,29 +2137,64 @@ func (engine *DockerTaskEngine) createContainer(task *apitask.Task, container *a
 	return metadata
 }
 
-func getFirelensLogConfig(task *apitask.Task, container *apicontainer.Container,
-	hostConfig *dockercontainer.HostConfig, cfg *config.Config) dockercontainer.LogConfig {
+// getFirelensLogConfig generates the fluentd log driver's log config.
+// Every container that wants to use Firelens for logging, gets one instance of the fluentd log driver associated to it.
+func getFirelensLogConfig(task *apitask.Task,
+	container *apicontainer.Container,
+	hostConfig *dockercontainer.HostConfig,
+	cfg *config.Config, dockerServerVersion string) (dockercontainer.LogConfig, error) {
+	var firelensConfig dockercontainer.LogConfig
+	// Set the log driver type
+	firelensConfig.Type = logDriverTypeFluentd
+	// Start from the existing container host config
+	hostConfigLogConfig := hostConfig.LogConfig
+	// Initialize a config to store the different log driver options
+	firelensConfig.Config = make(map[string]string)
+	// Generate a tag based on the task ID
 	fields := strings.Split(task.Arn, "/")
 	taskID := fields[len(fields)-1]
 	tag := fmt.Sprintf(fluentTagDockerFormat, container.Name, taskID)
+	firelensConfig.Config[logDriverTag] = tag
+	// Construct the fluent socket address
 	fluentd := socketPathPrefix + filepath.Join(cfg.DataDirOnHost, dataLogDriverPath, taskID, dataLogDriverSocketPath)
-	logConfig := hostConfig.LogConfig
-	bufferLimit, bufferLimitExists := logConfig.Config[apitask.FirelensLogDriverBufferLimitOption]
-	logConfig.Type = logDriverTypeFluentd
-	logConfig.Config = make(map[string]string)
-	logConfig.Config[logDriverTag] = tag
-	logConfig.Config[logDriverFluentdAddress] = fluentd
-	logConfig.Config[logDriverAsyncConnect] = strconv.FormatBool(cfg.FirelensAsyncEnabled.Enabled())
-	logConfig.Config[logDriverSubSecondPrecision] = strconv.FormatBool(true)
+	firelensConfig.Config[logDriverFluentdAddress] = fluentd
+	// Enable sub-second precision
+	firelensConfig.Config[logDriverSubSecondPrecision] = strconv.FormatBool(true)
+	// Set the log driver buffer limit if passed via the task payload
+	bufferLimit, bufferLimitExists := hostConfigLogConfig.Config[apitask.FirelensLogDriverBufferLimitOption]
 	if bufferLimitExists {
-		logConfig.Config[logDriverBufferLimit] = bufferLimit
+		firelensConfig.Config[logDriverBufferLimit] = bufferLimit
+	}
+	// Determine whether to use the "fluentd-async" option or the legacy "fluentd-async-connect" option.
+	// "fluentd-async-connect" option was deprecated in Docker v20.10.0 and removed in v28.0.0.
```

Comment:
Do we need a check and emit an error for >=v28 if `fluent-async-connect` is used?

Reply:
If Docker v20.10.0 and greater is being used, ECS agent will pass `fluentd-async` to Docker. If for some reason we are unable to determine the Docker server version, the task would see a CannotCreateContainer error.

Does the above answer your question?

Reply:
Ah, I was thinking that it was a decision that the customer would make. If we control it then I suspect we should be insulated from passing the deprecated value for later docker versions

Reply:
Yes, that's correct, the log config that agent passes to Docker is a map of key, value pairs. The key is controlled by the ECS agent (`fluentd-async` / `fluentd-async-connect`), and the value is configurable by the customer using the [`ECS_ENABLE_FIRELENS_ASYNC` agent config setting](https://github.com/aws/amazon-ecs-agent/blob/7dda21e06f907476e3cc45c2d510edabfc7893da/README.md?plain=1#L260).